### ISSUE TITLE
VARResults.irf_resim raises ValueError when exog variables are used

### ISSUE THREAD

**Issue by joonro** [statsmodels/statsmodels#6748]
#### Describe the bug

It seems `VARResults.irf_resim` raises `ValueError: endog and exog matrices are different sizes` when exog variables are used. Using the example from the [documentation](https://www.statsmodels.org/dev/vector_ar.html), it runs fine without any exog variables:

```python
# some example data
import numpy as np

import pandas

import statsmodels.api as sm

from statsmodels.tsa.api import VAR

mdata = sm.datasets.macrodata.load_pandas().data

# prepare the dates index
dates = mdata[['year', 'quarter']].astype(int).astype(str)

quarterly = dates["year"] + "Q" + dates["quarter"]

from statsmodels.tsa.base.datetools import dates_from_str

quarterly = dates_from_str(quarterly)

mdata = mdata[['realgdp','realcons','realinv']]

mdata.index = pandas.DatetimeIndex(quarterly)

data = np.log(mdata).diff().dropna()

# make a VAR model
model = VAR(data)

results = model.fit(maxlags=1)
irf_sim = results.irf_resim(orth=False, repl=100, steps=10, seed=1, burn=100, cum=False)
```

However, if I add an exog variable,

```python
data_exog = pandas.DataFrame(index=data.index)
data_exog['exovar1'] = np.random.normal(size=data_exog.shape[0])

# make a VAR model
model_with_exog = VAR(endog=data, exog=data_exog)

results_with_exog = model_with_exog.fit(maxlags=1)

irf_sim_with_exog = results_with_exog.irf_resim(orth=False, repl=100, steps=10, seed=1, burn=100, cum=False)

```

It raises the following `ValueError`:

<details>

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-4f525efaf429> in <module>
      7 results_with_exog = model_with_exog.fit(maxlags=1)
      8 
----> 9 irf_sim_with_exog = results_with_exog.irf_resim(orth=False, repl=100, steps=10, seed=1, burn=100, cum=False)

~\Anaconda3\lib\site-packages\pandas\util\_decorators.py in wrapper(*args, **kwargs)
    212                 else:
    213                     kwargs[new_arg_name] = new_arg_value
--> 214             return func(*args, **kwargs)
    215 
    216         return cast(F, wrapper)

~\Anaconda3\lib\site-packages\statsmodels\tsa\vector_ar\var_model.py in irf_resim(self, orth, repl, steps, seed, burn, cum)
   1612                               seed=seed, steps=nobs+burn)
   1613             sim = sim[burn:]
-> 1614             ma_coll[i, :, :, :] = fill_coll(sim)
   1615 
   1616         return ma_coll

~\Anaconda3\lib\site-packages\statsmodels\tsa\vector_ar\var_model.py in fill_coll(sim)
   1603 
   1604         def fill_coll(sim):
-> 1605             ret = VAR(sim, exog=self.exog).fit(maxlags=k_ar, trend=self.trend)
   1606             ret = ret.orth_ma_rep(maxn=steps) if orth else ret.ma_rep(maxn=steps)
   1607             return ret.cumsum(axis=0) if cum else ret

~\Anaconda3\lib\site-packages\statsmodels\tsa\vector_ar\var_model.py in __init__(self, endog, exog, dates, freq, missing)
    515     def __init__(self, endog, exog=None, dates=None, freq=None,
    516                  missing='none'):
--> 517         super(VAR, self).__init__(endog, exog, dates, freq, missing=missing)
    518         if self.endog.ndim == 1:
    519             raise ValueError("Only gave one variable to VAR")

~\Anaconda3\lib\site-packages\statsmodels\tsa\base\tsa_model.py in __init__(self, endog, exog, dates, freq, missing, **kwargs)
     45                  missing='none', **kwargs):
     46         super(TimeSeriesModel, self).__init__(endog, exog, missing=missing,
---> 47                                               **kwargs)
     48 
     49         # Date handling in indexes

~\Anaconda3\lib\site-packages\statsmodels\base\model.py in __init__(self, endog, exog, **kwargs)
    234 
    235     def __init__(self, endog, exog=None, **kwargs):
--> 236         super(LikelihoodModel, self).__init__(endog, exog, **kwargs)
    237         self.initialize()
    238 

~\Anaconda3\lib\site-packages\statsmodels\base\model.py in __init__(self, endog, exog, **kwargs)
     75         hasconst = kwargs.pop('hasconst', None)
     76         self.data = self._handle_data(endog, exog, missing, hasconst,
---> 77                                       **kwargs)
     78         self.k_constant = self.data.k_constant
     79         self.exog = self.data.exog

~\Anaconda3\lib\site-packages\statsmodels\base\model.py in _handle_data(self, endog, exog, missing, hasconst, **kwargs)
     98 
     99     def _handle_data(self, endog, exog, missing, hasconst, **kwargs):
--> 100         data = handle_data(endog, exog, missing, hasconst, **kwargs)
    101         # kwargs arrays could have changed, easier to just attach here
    102         for key in kwargs:

~\Anaconda3\lib\site-packages\statsmodels\base\data.py in handle_data(endog, exog, missing, hasconst, **kwargs)
    670     klass = handle_data_class_factory(endog, exog)
    671     return klass(endog, exog=exog, missing=missing, hasconst=hasconst,
--> 672                  **kwargs)

~\Anaconda3\lib\site-packages\statsmodels\base\data.py in __init__(self, endog, exog, missing, hasconst, **kwargs)
     86         self.k_constant = 0
     87         self._handle_constant(hasconst)
---> 88         self._check_integrity()
     89         self._cache = {}
     90 

~\Anaconda3\lib\site-packages\statsmodels\base\data.py in _check_integrity(self)
    433         if self.exog is not None:
    434             if len(self.exog) != len(self.endog):
--> 435                 raise ValueError("endog and exog matrices are different sizes")
    436 
    437     def wrap_output(self, obj, how='columns', names=None):

ValueError: endog and exog matrices are different sizes
```
</details>

When I check the values with a debugger, in `statsmodels\tsa\vector_ar\var_model.py(1614)`, `sim` has 1 less rows than `self.exog`:

```python
> c:\users\joon\anaconda3\lib\site-packages\statsmodels\tsa\vector_ar\var_model.py(1614)irf_resim()
   1612                               seed=seed, steps=nobs+burn)
   1613             sim = sim[burn:]
-> 1614             ma_coll[i, :, :, :] = fill_coll(sim)
   1615 
   1616         return ma_coll
```

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

<details>

[paste the output of ``import statsmodels.api as sm; sm.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
Python: 3.7.3.final.0

statsmodels
===========

Installed: 0.11.1 (C:\Users\joon\Anaconda3\lib\site-packages\statsmodels)

Required Dependencies
=====================

cython: 0.29.17 (C:\Users\joon\Anaconda3\lib\site-packages\Cython)
numpy: 1.18.1 (C:\Users\joon\Anaconda3\lib\site-packages\numpy)
scipy: 1.4.1 (C:\Users\joon\Anaconda3\lib\site-packages\scipy)
pandas: 1.0.1 (C:\Users\joon\Anaconda3\lib\site-packages\pandas)
    dateutil: 2.8.1 (C:\Users\joon\Anaconda3\lib\site-packages\dateutil)
patsy: 0.5.1 (C:\Users\joon\Anaconda3\lib\site-packages\patsy)

Optional Dependencies
=====================

matplotlib: 3.1.3 (C:\Users\joon\Anaconda3\lib\site-packages\matplotlib)
    backend: module://ipykernel.pylab.backend_inline 
cvxopt: Not installed
joblib: 0.15.1 (C:\Users\joon\Anaconda3\lib\site-packages\joblib)

Developer Tools
================

IPython: 7.13.0 (C:\Users\joon\Anaconda3\lib\site-packages\IPython)
    jinja2: 2.11.2 (C:\Users\joon\Anaconda3\lib\site-packages\jinja2)
sphinx: 3.0.3 (C:\Users\joon\Anaconda3\lib\site-packages\sphinx)
    pygments: 2.6.1 (C:\Users\joon\Anaconda3\lib\site-packages\pygments)
pytest: 5.4.2 (C:\Users\joon\Anaconda3\lib\site-packages\pytest)
virtualenv: Not installed

</details>

**Comment by joonro** [comment]
It seems the issues is that in the following code, it is asking for `steps=nobs+burn` (L1621), but `nobs` is the number of observations after taking out the lag, which is 1 in this example (`maxlags=1`). I think it should instead ask for `steps=nobs+k_ar+burn`, because in the L1615, it is passing `sim` as the `endog` variables with `maxlags=k_ar`. Please let me know if this is correct, I'm happy to open a PR.

https://github.com/statsmodels/statsmodels/blob/1fa83517280a03057689f24ad5d712ee894279a9/statsmodels/tsa/vector_ar/var_model.py#L1614-L1624

**Comment by josef-pkt** [comment]
You can prepare and open a PR.
It will take some time to check this.

One thing to verify is whether the attached exog are still the original shape or some truncated shape. I don't remember the details but there were too many versions of the data attached.

AFAIK, exog hasn't been well tested for VAR and was mainly added to support (deterministic exog in) VECM.

**Comment by josef-pkt** [comment]
maybe related #4542

**Comment by joonro** [comment]
Thanks for the quick reply!

> One thing to verify is whether the attached exog are still the original shape or some truncated shape. I don't remember the details but there were too many versions of the data attached.

Yes `exog` is not touched - in L1605, it is just passing `self.exog` to `VAR`, and `self.exog` is the original argument that was passed in initializing the VAR object it seems.

**Comment by josef-pkt** [comment]
line 1621

is self.nobs the nobs of the original data or the nobs without initial observations (nobs_original - k_ar)?

i.e. is nobs == self.exog.shape[0] ?

**Comment by joonro** [comment]
> line 1621
> 
> is self.nobs the nobs of the original data or the nobs without initial observations (nobs_original - k_ar)?
> 
> i.e. is nobs == self.exog.shape[0] ?

`nobs` is nobs without initial observations (`nobs_original - k_ar`). In my debugging session in `irf_resim` stack) of the same example), 

```python
In[12]: self.nobs
Out[12]: 201
In[13]: self.endog.shape[0]
Out[13]: 202
```

**Comment by josef-pkt** [comment]
ok,
to make the code self documenting

add a new local variable `nobs_original = nobs + k_ar` and use that as `steps`
and then see whether that resolves it.

This means that we simulate a shorter time series than the original in the current code.

There is also a `VARProcess.simulate_var` but AFAICS it's not used for any computation in the model.

**Comment by joonro** [comment]
Yes the change (`nobs_original = nobs + k_ar` and then passing `steps=nobs_original+burn`) resolves it.

**Comment by josef-pkt** [comment]
good, do the results look ok?
We might need just a regression test in the unit test to make sure it runs. It won't be easy to verify with other packages that the numbers are correct.
(regression test means using current result as the values to be asserted)


just checking:
util.varsim simulates without exog. It adds intercept but that is here not used as an offset (as in GLM to add a term with fixed coefficient 1)

So `irf_resim` simulate a model where coefficient for exog are all zero, but then estimates the model with exog.
Based on what I know in other contexts, this works for confidence interval and hypothesis testing because the asymptotic distribution is invariant to the values of the exog parameters. (and exog are strictly exogenous.)

In contrast, `VARProcess.simulate_var` simulates conditional on the effect of the exog and trend, using the intercept as an offset.

**Comment by joonro** [comment]
> good, do the results look ok?




> We might need just a regression test in the unit test to make sure it runs. It won't be easy to verify with other packages that the numbers are correct.
> (regression test means using current result as the values to be asserted)
> 
> just checking:
> util.varsim simulates without exog. It adds intercept but that is here not used as an offset (as in GLM to add a term with fixed coefficient 1)
> 
> So `irf_resim` simulate a model where coefficient for exog are all zero, but then estimates the model with exog.
> Based on what I know in other contexts, this works for confidence interval and hypothesis testing because the asymptotic distribution is invariant to the values of the exog parameters. (and exog are strictly exogenous.)
> 
> In contrast, `VARProcess.simulate_var` simulates conditional on the effect of the exog and trend, using the intercept as an offset.

**Comment by joonro** [comment]
> good, do the results look ok?

When I checked the results, I found another issue - `varsim` sets `np.random.RandomState` given the `seed` argument. But in `irf_resim` it calls `varsim` with the fixed `seed` argument in the loop - so actually `varsim` keeps returning the same values.

https://github.com/statsmodels/statsmodels/blob/1fa83517280a03057689f24ad5d712ee894279a9/statsmodels/tsa/vector_ar/var_model.py#L1619-L1624

https://github.com/statsmodels/statsmodels/blob/b66e92af24296b195f2e556367e56ddb98bebd60/statsmodels/tsa/vector_ar/util.py#L225

> Based on what I know in other contexts, this works for confidence interval and hypothesis testing because the asymptotic distribution is invariant to the values of the exog parameters. (and exog are strictly exogenous.)
> 
> In contrast, `VARProcess.simulate_var` simulates conditional on the effect of the exog and trend, using the intercept as an offset.

I see - thanks for the heads-up. I'm not 100% sure, but it seems that behavior is correct, since with `irf` one is assuming that no exog variables are changing. I need to double check to make sure.

**Comment by josef-pkt** [comment]
> But in irf_resim it calls varsim with the fixed seed argument in the loop - so actually varsim keeps returning the same values.

This only happens if you set a seed not equal to None, doesn't it?

It's a bug, but a bit surprising that we never found it. Maybe, there isn't much use of it with a specified seed.

**Comment by joonro** [comment]
I think still you want to be able to fix the seed for `irf_resim` for replication standpoint. If setting `np.random.RandomState` to be `None` does not change the seed (I'm not sure how it behaves), then would it be okay to set it once in `irf_resim` before starting the loop, and passing `seed=None` when it calls `varsim` inside loop?

**Comment by josef-pkt** [comment]
I think we should clean up the seed handling properly and create a RandomState in irf_resim and use that in loops and varsim,
Other packages like scikit learn and scipy have the proper helper function for setting random state.

I'm busy right now, so I cannot look for it until later.

**Comment by josef-pkt** [comment]
here is a PR that changed this https://github.com/statsmodels/statsmodels/pull/3043/files
old issue for proper random state and seed handling #306
here is a version in one of my unmerged PRs but already a few years old https://github.com/josef-pkt/statsmodels/commit/0fd2fd5fddf6c00d0cf3c09e42ce31615a0140ef

here is the current scipy version https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L177
and sklearn version https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L852

we need to avoid that any of the functions that are called set the np.random.seed
this is a bug ( I guess) introduced in the above PR, before we didn't set seed if seed is None
https://github.com/statsmodels/statsmodels/pull/3043/files#diff-df55526cf29db3929289fd16fc3f3393L189

If seed is None, then it reinitializes the random state and does not continue with the current global stats, so we need to avoid this default treatment of `seed=None`. That's what the `check_random_state` also does.
" If seed is None, then the MT19937 BitGenerator is initialized by reading data from /dev/urandom (or the Windows analogue) if available or seed from the clock otherwise."
https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState

**Comment by josef-pkt** [comment]
I think we can do something similar to https://github.com/josef-pkt/statsmodels/commit/0fd2fd5fddf6c00d0cf3c09e42ce31615a0140ef
but put the check function somewhere in tools.

@joonro  You could either fix both problems in one PR, or you fix the nobs shape error in one PR and then we add the check random state handling for seed in a second PR.

**Comment by joonro** [comment]
Got it. Thank you very much for the follow up. I think I will create two separate PRs.
